core/common/src/internal/format/parser/ParserOperation.kt

```diff
@@ -139,6 +139,130 @@ internal class UnconditionalModification<Output>(
     }
 }
 
+internal class TimeZoneParserOperation<Output>(
+    private val setter: AssignableField<Output, String>
+) : ParserOperation<Output> {
+
+    override fun consume(storage: Output, input: CharSequence, startIndex: Int): ParseResult {
+        if (startIndex >= input.length) return ParseResult.Error(startIndex) { "Unexpected end of input" }
+        val lastMatch = validateTimezone(input, startIndex)
+        if (lastMatch > startIndex) {
+            setter.setWithoutReassigning(storage, input.substring(startIndex, lastMatch), startIndex, lastMatch)
+            return ParseResult.Ok(lastMatch)
+        }
+        return ParseResult.Error(startIndex) { "Invalid timezone format" }
+    }
+
+    companion object {
+        enum class State {
+            START,
+            AFTER_PREFIX,
+            AFTER_SIGN,
+            AFTER_HOUR,
+            AFTER_HOUR_COLON,
+            AFTER_MINUTE,
+            AFTER_MINUTE_COLON,
+            END,
+            INVALID
+        }
+
+        private fun validateTimezone(input: CharSequence, startIndex: Int): Int {
+            var index = startIndex
+            var lastValidIndex = startIndex
+
+            fun hasEnoughChars(length: Int) = index + length <= input.length
+
+            fun validatePrefix(length: Int, validValues: List<String>): Boolean {
+                if (hasEnoughChars(length) && input.substring(index, index + length) in validValues) {
+                    index += length
+                    lastValidIndex = index
+                    return true
+                }
+                return false
+            }
+
+            fun validateTimeComponent(length: Int): Boolean {
+                if (hasEnoughChars(length) && input.slice(index..<(index + length)).all { it.isDigit() }) {
+                    index += length
+                    lastValidIndex = index
+                    return true
+                }
+                return false
+            }
+
+            var state = State.START
+            while (index < input.length) {
```

Comment:
This state machine accepts `+12:3456` and `+1234:56`, none of which is in a format that `TimeZone.of` accepts.